File: python/thunder_rpc/pool.py
"""Optional connection pool (CLT-080) — a layer **above** the
single-connection clients (CLT-001: "pooling is a layer above").

Under a mandatory-``HELLO`` profile with ``auth_required``, a fresh connection
costs a handshake round trip before the first request. A caller that opens a
connection per operation therefore pays that round trip every time. The pool
amortizes it: ``N`` operations over a checked-out connection pay **one**
connect and **one** handshake, not ``N``.

The shape is deliberately minimal — a fixed number of connections bounded by a
semaphore, an idle list, lazy connect on first checkout, and a guard that
returns the connection to the pool on exit. It is **not** an external pool
library. Health checks, background reaping and min-idle warmup are out of
scope; a poisoned connection (CLT-014) is dropped on return and the next
checkout connects fresh, leaving reconnect to CLT-030 rather than the pool.

The pool adds **no wire behavior**: it builds the same client as
``Client.connect`` / ``AsyncClient.connect`` from a :class:`Config` and a
:class:`ClientConfig`. ``max_in_flight`` (CLT-012) stays a per-connection
bound; the pool bounds connections, not in-flight calls.

Python idiom: the guard is a context manager (FR-28) — ``with pool.acquire()
as client:`` for the sync :class:`Pool`, ``async with pool.acquire() as
client:`` for the async :class:`AsyncPool` — releasing the connection on exit,
including the error path.

Example::

    from thunder_rpc import ClientConfig, Config, Pool

    app = Config.standard().with_scheme("myapp").with_port(9000)
    pool = Pool("myapp://localhost", app, ClientConfig(), max_connections=8)
    with pool.acquire() as client:      # reuses an idle connection, or dials one
        pong = client.call("PING")
        assert pong.as_str() == "PONG"
    # the connection returns to the pool when the `with` block exits.
"""

from __future__ import annotations

import asyncio
import threading

from .aio import AsyncClient
from .client import Client
from .client_config import ClientConfig
from .config import Config


class Pool:
    """A bounded pool of synchronous :class:`~thunder_rpc.client.Client`\\ s
    over one endpoint (CLT-080).

    At most ``max_connections`` connections are live at once; a checkout beyond
    that awaits a return. Connections are dialed lazily — construction opens
    none — and reused across checkouts so the handshake is paid once per
    connection, not once per operation. ``max_connections`` is clamped to at
    least 1.
    """

    def __init__(
        self,
        endpoint: str,
        config: Config,
        client_config: ClientConfig | None = None,
        max_connections: int = 8,
    ) -> None:
        self._endpoint = endpoint
        self._config = config
        self._client_config = (
            client_config if client_config is not None else ClientConfig()
        )
        max_conn = max(1, max_connections)
        #: Bounds live + checked-out connections to ``max_connections``.
        self._permits = threading.BoundedSemaphore(max_conn)
        self._idle: list[Client] = []
        self._idle_lock = threading.Lock()

    def acquire(self) -> _PooledConn:
        """Return a guard; entering its ``with`` block checks out a connection
        (reusing an idle, live one or dialing a fresh one) and yields the
        client, returning it to the pool on exit (CLT-080)."""
        return _PooledConn(self)

    def idle_count(self) -> int:
        """Idle connections currently parked in the pool. For diagnostics and
        tests — production code should not branch on it."""
        with self._idle_lock:
            return len(self._idle)

    def _checkout(self) -> Client:
        # Awaits a return when max_connections are already checked out.
        self._permits.acquire()
        try:
            # Reuse the newest idle connection that is still live; discard any
            # poisoned (CLT-014) while sitting idle.
            reused: Client | None = None
            with self._idle_lock:
                while self._idle:
                    candidate = self._idle.pop()
                    if candidate.is_alive():
                        reused = candidate
                        break
                    candidate.close()
            if reused is not None:
                return reused
            return Client.connect(self._endpoint, self._config, self._client_config)
        except BaseException:
            self._permits.release()
            raise

    def _checkin(self, client: Client) -> None:
        # CLT-014: only a live connection returns to the pool; a poisoned or
        # closed one is dropped here, and the next checkout dials fresh.
        try:
            if client.is_alive():
                with self._idle_lock:
                    self._idle.append(client)
            else:
                client.close()
        finally:
            self._permits.release()


class _PooledConn:
    """Guard from :meth:`Pool.acquire`. A context manager: entering checks out
    a client, exiting returns it to the pool (or drops it if poisoned,
    CLT-014). ``__exit__`` runs on the error path too, so a connection never
    leaks."""

    __slots__ = ("_pool", "_client")

    def __init__(self, pool: Pool) -> None:
        self._pool = pool
        self._client: Client | None = None

    def __enter__(self) -> Client:
        self._client = self._pool._checkout()
        return self._client

    def __exit__(self, *_exc: object) -> None:
        client = self._client
        self._client = None
        if client is not None:
            self._pool._checkin(client)


class AsyncPool:
    """A bounded pool of :class:`~thunder_rpc.aio.AsyncClient`\\ s over one
    endpoint (CLT-080) — the asyncio mirror of :class:`Pool`, identical
    semantics, ``async with`` idiom only (FR-28)."""

    def __init__(
        self,
        endpoint: str,
        config: Config,
        client_config: ClientConfig | None = None,
        max_connections: int = 8,
    ) -> None:
        self._endpoint = endpoint
        self._config = config
        self._client_config = (
            client_config if client_config is not None else ClientConfig()
        )
        self._max = max(1, max_connections)
        self._permits: asyncio.BoundedSemaphore | None = None
        self._idle: list[AsyncClient] = []

    def acquire(self) -> _AsyncPooledConn:
        """Return a guard; entering its ``async with`` block checks out a
        connection (reusing an idle, live one or dialing a fresh one) and
        yields the client, returning it to the pool on exit (CLT-080)."""
        return _AsyncPooledConn(self)

    def idle_count(self) -> int:
        """Idle connections currently parked in the pool. For diagnostics and
        tests — production code should not branch on it."""
        return len(self._idle)

    def _semaphore(self) -> asyncio.BoundedSemaphore:
        # Created lazily so the pool can be constructed off the event loop.
        if self._permits is None:
            self._permits = asyncio.BoundedSemaphore(self._max)
        return self._permits

    async def _checkout(self) -> AsyncClient:
        permits = self._semaphore()
        # Awaits a return when max_connections are already checked out.
        await permits.acquire()
        try:
            # Reuse the newest idle, live connection; discard poisoned ones.
            reused: AsyncClient | None = None
            while self._idle:
                candidate = self._idle.pop()
                if candidate.is_alive():
                    reused = candidate
                    break
                await candidate.close()
            if reused is not None:
                return reused
            return await AsyncClient.connect(
                self._endpoint, self._config, self._client_config
            )
        except BaseException:
            permits.release()
            raise

    async def _checkin(self, client: AsyncClient) -> None:
        # CLT-014: only a live connection returns to the pool; a poisoned or
        # closed one is dropped here, and the next checkout dials fresh.
        try:
            if client.is_alive():
                self._idle.append(client)
            else:
                await client.close()
        finally:
            self._semaphore().release()


class _AsyncPooledConn:
    """Guard from :meth:`AsyncPool.acquire`. An async context manager:
    entering checks out a client, exiting returns it to the pool (or drops it
    if poisoned, CLT-014). ``__aexit__`` runs on the error path too."""

    __slots__ = ("_pool", "_client")

    def __init__(self, pool: AsyncPool) -> None:
        self._pool = pool
        self._client: AsyncClient | None = None

    async def __aenter__(self) -> AsyncClient:
        self._client = await self._pool._checkout()
        return self._client

    async def __aexit__(self, *_exc: object) -> None:
        client = self._client
        self._client = None
        if client is not None:
            await self._pool._checkin(client)


__all__ = ["AsyncPool", "Pool"]

File: csharp/HiveLLM.Thunder/Pool.cs
namespace HiveLLM.Thunder;

/// <summary>
/// Optional connection pool (CLT-080) — a layer <b>above</b> the
/// single-connection <see cref="ThunderClient"/> (CLT-001: "pooling is a layer
/// above").
/// <para>
/// Under a mandatory-<c>HELLO</c> profile a fresh connection costs a handshake
/// round trip before the first request; a caller that opens a connection per
/// operation pays that every time. The pool amortizes it: <c>N</c> operations
/// over a checked-out connection pay <b>one</b> connect and <b>one</b>
/// handshake, not <c>N</c>.
/// </para>
/// <para>
/// The shape is deliberately minimal — a fixed number of connections bounded by
/// a <see cref="SemaphoreSlim"/>, an idle stack, lazy connect on first checkout,
/// and a guard (<see cref="PooledConn"/>) that returns the connection on
/// <see cref="PooledConn.DisposeAsync"/>. It is <b>not</b> an external pool
/// library: health checks, background reaping and min-idle warmup are out of
/// scope; a poisoned connection (CLT-014) is dropped on return and the next
/// checkout connects fresh, leaving reconnect to CLT-030 rather than the pool.
/// </para>
/// <para>
/// The pool adds <b>no wire behavior</b>: it builds the same
/// <see cref="ThunderClient"/> as <see cref="ThunderClient.ConnectAsync"/> from
/// a <see cref="Config"/> and <see cref="ClientConfig"/>, and the
/// single-connection client's API is unchanged.
/// <see cref="Config.MaxInFlight"/> (CLT-012) stays a per-connection bound; the
/// pool bounds connections, not in-flight calls.
/// </para>
/// <example>
/// <code>
/// var app = Config.Standard() with { Scheme = "myapp", DefaultPort = 9000 };
/// using var pool = new Pool("myapp://localhost", app, new ClientConfig(), 8);
/// await using var conn = await pool.AcquireAsync(); // reuses idle, or dials
/// var pong = await conn.CallAsync("PING");
/// // `conn` returns the connection to the pool when it is disposed.
/// </code>
/// </example>
/// </summary>
public sealed class Pool : IDisposable
{
    private readonly string _endpoint;
    private readonly Config _config;
    private readonly ClientConfig _clientConfig;

    /// <summary>Bounds live + checked-out connections to <see cref="MaxConnections"/>.</summary>
    private readonly SemaphoreSlim _permits;

    private readonly object _idleLock = new();

    /// <summary>Idle connections available for reuse — newest reused first (LIFO).</summary>
    private readonly Stack<ThunderClient> _idle;

    private bool _disposed;

    /// <summary>
    /// Build a pool for <paramref name="endpoint"/>. Opens no connections — the
    /// first <see cref="AcquireAsync"/> dials the first one.
    /// <paramref name="maxConnections"/> is clamped to at least 1.
    /// </summary>
    /// <param name="endpoint">The endpoint every pooled client dials (CLT-070).</param>
    /// <param name="config">The application's protocol config — the dialect (PRO-001).</param>
    /// <param name="clientConfig">This caller's credentials/timeouts, shared by every pooled client.</param>
    /// <param name="maxConnections">The fixed connection bound; clamped to <c>&gt;= 1</c>.</param>
    public Pool(string endpoint, Config config, ClientConfig clientConfig, int maxConnections)
    {
        ArgumentNullException.ThrowIfNull(endpoint);
        ArgumentNullException.ThrowIfNull(config);
        ArgumentNullException.ThrowIfNull(clientConfig);
        var max = Math.Max(1, maxConnections);
        _endpoint = endpoint;
        _config = config;
        _clientConfig = clientConfig;
        MaxConnections = max;
        _permits = new SemaphoreSlim(max, max);
        _idle = new Stack<ThunderClient>(max);
    }

    /// <summary>The fixed connection bound (clamped to <c>&gt;= 1</c>).</summary>
    public int MaxConnections { get; }

    /// <summary>
    /// Idle connections currently parked in the pool. For diagnostics and tests
    /// — production code should not branch on it.
    /// </summary>
    public int IdleCount
    {
        get
        {
            lock (_idleLock)
            {
                return _idle.Count;
            }
        }
    }

    /// <summary>
    /// Check out a connection. Reuses an idle, <b>live</b> connection when one
    /// is available; otherwise dials and handshakes a fresh one (CLT-002).
    /// Awaits a return when <see cref="MaxConnections"/> are already checked out.
    /// The returned <see cref="PooledConn"/> returns the connection to the pool
    /// when disposed.
    /// </summary>
    /// <param name="cancellationToken">Cancels the wait for a slot and the dial.</param>
    /// <exception cref="ThunderConnectionException">The dial failed, or the pool is disposed.</exception>
    /// <exception cref="ThunderTimeoutException">The connect timeout elapsed (CLT-001).</exception>
    /// <exception cref="ThunderAuthException">The server rejected the handshake (CLT-003).</exception>
    public async Task<PooledConn> AcquireAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            await _permits.WaitAsync(cancellationToken).ConfigureAwait(false);
        }
        catch (ObjectDisposedException)
        {
            throw new ThunderConnectionException("connection pool is closed");
        }

        try
        {
            // Reuse the newest idle connection that is still live; discard any
            // that were poisoned (CLT-014) while sitting idle.
            ThunderClient? reused = null;
            lock (_idleLock)
            {
                while (_idle.Count > 0)
                {
                    var candidate = _idle.Pop();
                    if (candidate.IsAlive)
                    {
                        reused = candidate;
                        break;
                    }

                    candidate.Dispose();
                }
            }

            var client = reused ?? await ThunderClient
                .ConnectAsync(_endpoint, _config, _clientConfig, cancellationToken)
                .ConfigureAwait(false);
            return new PooledConn(this, client);
        }
        catch
        {
            // The checkout failed to hand out a guard — release the slot the
            // guard would otherwise have held.
            _permits.Release();
            throw;
        }
    }

    /// <summary>
    /// Return a checked-out connection (called by <see cref="PooledConn"/> on
    /// dispose). A live connection is parked for reuse; a poisoned or closed one
    /// is dropped and the next checkout dials fresh (CLT-014/030). The slot is
    /// released either way.
    /// </summary>
    internal void Return(ThunderClient client)
    {
        var parked = false;
        if (!_disposed && client.IsAlive)
        {
            lock (_idleLock)
            {
                if (!_disposed)
                {
                    // Park before releasing the slot so a waiter wakes to find
                    // this connection idle rather than dialing a redundant one.
                    _idle.Push(client);
                    parked = true;
                }
            }
        }

        if (!parked)
        {
            client.Dispose();
        }

        try
        {
            _permits.Release();
        }
        catch (ObjectDisposedException)
        {
            // The pool was disposed while this connection was checked out; the
            // slot no longer exists to release.
        }
    }

    /// <summary>
    /// Dispose the pool: close every idle connection. Checked-out connections
    /// close normally when their guard is disposed.
    /// </summary>
    public void Dispose()
    {
        List<ThunderClient> idle;
        lock (_idleLock)
        {
            if (_disposed)
            {
                return;
            }

            _disposed = true;
            idle = new List<ThunderClient>(_idle);
            _idle.Clear();
        }

        foreach (var client in idle)
        {
            client.Dispose();
        }

        _permits.Dispose();
    }
}

/// <summary>
/// Guard from <see cref="Pool.AcquireAsync"/>. Exposes the checked-out
/// <see cref="ThunderClient"/> and returns the connection to the pool on
/// <see cref="DisposeAsync"/> so the next checkout reuses it — unless the
/// connection was poisoned, in which case it is dropped and the next checkout
/// connects fresh (CLT-014/030). The .NET idiom for the Rust RAII guard:
/// <c>await using var conn = await pool.AcquireAsync();</c>.
/// </summary>
public sealed class PooledConn : IAsyncDisposable, IDisposable
{
    private readonly Pool _pool;
    private ThunderClient? _client;

    internal PooledConn(Pool pool, ThunderClient client)
    {
        _pool = pool;
        _client = client;
    }

    /// <summary>
    /// The checked-out client. Valid until this guard is disposed.
    /// </summary>
    /// <exception cref="ObjectDisposedException">The guard has been disposed.</exception>
    public ThunderClient Client =>
        _client ?? throw new ObjectDisposedException(nameof(PooledConn));

    /// <summary>Issue one call on the checked-out connection (CLT-020).</summary>
    public Task<Value> CallAsync(
        string command,
        IReadOnlyList<Value>? args = null,
        CancellationToken cancellationToken = default) =>
        Client.CallAsync(command, args, cancellationToken);

    /// <summary>Return the connection to the pool (or drop it if poisoned). Idempotent.</summary>
    public ValueTask DisposeAsync()
    {
        ReturnToPool();
        return ValueTask.CompletedTask;
    }

    /// <inheritdoc cref="DisposeAsync" />
    public void Dispose() => ReturnToPool();

    private void ReturnToPool()
    {
        var client = Interlocked.Exchange(ref _client, null);
        if (client is not null)
        {
            _pool.Return(client);
        }
    }
}
